src/radical/pilot/raptor/master.py

```diff
@@ -271,67 +332,76 @@ def _state_cb(self, topic, msg):
     #
-    def submit_workers(self, descr: Dict[str, Union[str, int]],
-                             count: int) -> None:
+    def worker_state_cb(self, worker_dict, state):
         '''
-        Submit`count` workers per given `descr`, and pass the queue info as
-        configuration file.  Do *not* wait for the workers to come up - they are
-        expected to register via the control channel.
+        This callback can be overloaded - it will be invoked whenever the master
+        receives a state update for a worker it is connected to.
+        '''
+
+        pass
 
-        The `descr` dict is expected to support the following keys:
 
-          - named_env       : environment to use (same as master usually)
-          - ranks           : number of MPI ranks per worker
-          - cores_per_rank  : int, number of cores per worker rank
-          - gpus_per_rank   : int, number of gpus per worker rank
-          - worker_class    : str, type name of worker class to execute
-          - worker_file     : str, optional if an RP worker class is used
+    # --------------------------------------------------------------------------
+    #
+    def submit_workers(self, descriptions: List[TaskDescription]
+                      ) -> List[str]:
+        '''
+        Submit a raptor workers per given `descriptions` element and pass the
+        queue raptor info as configuration file.  Do *not* wait for the workers
+        to come up - they are expected to register via the control channel.
 
-        Note that only one rank (presumably rank 0) should register with the
-        master - the worker ranks are expected to syncronize their ranks as
+        The task `descriptions` specifically support the following keys:
+
+          - raptor_class: str, type name of worker class to execute
+          - raptor_file : str, optional if an RP worker class is used
+
+        Note that only one worker rank (presumably rank 0) should register with
+        the master - the workers are expected to syncronize their ranks as
         needed.
         '''
+
+        # FIXME registry: use registry instead of config files
+
         tasks = list()
-        base  = self._uid
+        for td in descriptions:
 
-        worker_file  = descr.pop('worker_file', '')
-        worker_class = descr.pop('worker_class', 'DefaultWorker')
+            # ensure that defaults and backward compatibility kick in
+            td.verify()
 
-        td = TaskDescription(descr)
-        td.mode       = RAPTOR_WORKER
-        td.executable = 'radical-pilot-raptor-worker'
+            # sharing GPUs among multiple ranks not yet supported
+            if td.gpus_per_rank and not td.gpus_per_rank.is_integer():
+                raise RuntimeError('GPU sharing for workers is not supported')
 
-        # ensure that defaults and backward compatibility kick in
-        td.verify()
+            td.mode      = RAPTOR_WORKER
```

Comment:
The mode is assigned by the `submit_workers` call rather than by the user?

Reply:
At this point in the code we consider it safe to assume that the description targets a worker, so we set it to make sure the downstream code knows what to do.  Consider it a service to the end user :-)

Reply:
I changed this now to a check.